### PR TITLE
fix: more strict system32 path check

### PR DESCRIPTION
**Pull request by ThomasKoscheck**
# PR Summary

Constrained Language Mode enforced with Applocker or Environment variable `__PSLockdownPolicy` can be bypassed by naming the script to "system32.ps1" because of sloppy path check

Info: I reported this as https://msrc.microsoft.com/report/vulnerability/VULN-093434, but the case was closed there without further investigation

## PR Context

To bypass the constrained language mode the path from where your script is being executed, needs to contain the string system32, meaning even if you rename the script to `system32.ps1`.

From a security perspective, this is of course a nightmare!

Steps to Reproduce (Add details for how we can reproduce the issue)
Activate Constrained Language mode via Applocker policy, the `__PSLockdownPolicy` variable or set it with 
`$ExecutionContext.SessionState.LanguageMode = "Constrained"` for current PowerShell session.

Execute a PowerShell script with the following content to display the current language mode

`$ExecutionContext.SessionState.LanguageMode`
This should output "ConstrainedLanguage" if step 1 was applied successfully.

Rename PowerShell script from step 2 to system32.ps1

The output will be "FullLanguage", as we have bypassed the security check

## More details

https://www.blackhillsinfosec.com/constrained-language-mode-bypass-when-pslockdownpolicy-is-used/

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by PaulHigin** [comment]
The `system32` file name check applies only when there is no WDAC or AppLocker system lock down policy being enforced, and is used with the `__PSLockdownPolicy` environment variable testing/debugging hook.  The `__PSLockdownPolicy` var and `system32` name check is for testing only and cannot override a policy lock down enforcement.  

Constrained language mode is only useful as a security measure when enforced by system policy.  Otherwise it is easily bypassed.

/cc: @TravisEz13

**Comment by ThomasKoscheck** [comment]
Hi @PaulHigin,

I I thought so too, but it works with Applocker enforced Constrained Language Mode as well.

**Comment by TravisEz13** [comment]
@ThomasKoscheck 
I cannot reproduce your issue with the default AppLocker Rules.  Please provide the AppLocker rules you used to reproduce the issue.
image

File: src/System.Management.Automation/security/wldpNativeMethods.cs [hunk 1]
```diff
@@ -494,9 +494,8 @@ private static SystemEnforcementMode GetDebugLockdownPolicy(string path)
             // Support fall-back debug hook for path exclusions on non-WOA platforms
             if (path != null)
             {
-                // Assume everything under SYSTEM32 is trusted, with a purposefully sloppy
-                // check so that we can actually put it in the filename during testing.
-                if (path.Contains("System32", StringComparison.OrdinalIgnoreCase))
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/security/wldpNativeMethods.cs:499]: This check is weak, but it appears to only be used for test cases and not exploitable beyond that

**Comment by TravisEz13** [comment]
Here is a more complete repro.

I setup AppLocker to actually enable System Lockdown, using the following rules:
image
I force applied Group Policy and restarted the AppLocker service, then ran this:
image


I retried the repro running this,
`[Environment]::SetEnvironmentVariable(‘__PSLockdownPolicy‘, ‘4’, ‘Machine‘)`
from an elevated PowerShell session before starting, with no change in the result.

I believe there was some misconfiguration of AppLocker policy.  

`__PSLockdownPolicy` is only meant for testing and it is acceptable when running with only that policy, that we allow any file that passes the rule in the existing code.

**Comment by ThomasKoscheck** [comment]
Hi, 
I created a test ad lab to reproduce with GPOs and a clean applocker policy.

Just as @TravisEz13 had already shown, I could not reproduce the error with the fresh applocker policy.

So, the issue can be closed, sorry for my mistake, and thanks for looking at it!

**Pull request closed without merging**